### PR TITLE
Implementation of new Time of Use Tariff controller

### PR DESCRIPTION
**Pull request by venu-sagar**
A new Time of Use tariff controller has been implemented, replacing the previous controller called TimeOfUseTariffDischarge. 

The new controller will have a new Factory id: 'Controller.Ess.Time-Of-Use-Tariff' replacing the old controller factory Id: 'Controller.Ess.Time-Of-Use-Tariff.Discharge'.

->To accommodate different markets with dynamic pricing, we have added control modes:

    - CHARGE_CONSUMPTION: This mode is suitable for markets where charging from the grid is allowed and preferred.

    - DELAY_DISCHARGE: This mode is designed for markets where charging from the grid is restricted, and the discharge is delayed accordingly.

-> Added Jsonrpc request handler to show future data in UI.
-> Modified Awattar provider implementation to handle querying for a price every hour instead of once a day.
-> Added appropriate channels for better debugging.
-> Added suitable test cases.

### CODE REVIEW

File: io.openems.edge.timeofusetariff.awattar/src/io/openems/edge/timeofusetariff/awattar/TimeOfUseTariffAwattarImpl.java [hunk 1]
```diff
@@ -115,12 +115,13 @@ protected void deactivate() {
...
-		 * Schedule next price update for 2 pm
+		 * Schedule next price update every hour
 		 */
 		var now = ZonedDateTime.now();
-		var nextRun = now.withHour(14).truncatedTo(ChronoUnit.HOURS);
+		var nextRun = now.plusHours(1).truncatedTo(ChronoUnit.HOURS);
+
 		if (now.isAfter(nextRun)) {
-			nextRun = nextRun.plusDays(1);
+			nextRun = nextRun.plusHours(1);
```
Comment (sfeilmeier) [inline io.openems.edge.timeofusetariff.awattar/src/io/openems/edge/timeofusetariff/awattar/TimeOfUseTariffAwattarImpl.java:124]: Why is that required? The prices won't change anymore after 2pm, right? We might even meet API (fair use) restrictions if we query to often.

Comment (venu-sagar) [reply io.openems.edge.timeofusetariff.awattar/src/io/openems/edge/timeofusetariff/awattar/TimeOfUseTariffAwattarImpl.java:124]: This is to ensure a complete schedule for the entire day. Previously, prices were queried at 2 pm, providing 24 values for the next 24 hours. that means at 1:00 pm the next day, we have only one-hour price and the schedule is only done till that hour. 

One alternative way is to query once at 2:00 pm and again at midnight (12:00 am) as well to get prices for the whole day, so that schedule is done for the whole day.

File: io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/AwattarHourly.java [hunk 2]
```diff
@@ -66,34 +96,19 @@ public AwattarHourly(@Reference ComponentManager componentManager, ComponentCont
...
-					new EdgeConfig.Component(timeOfUseTariffId, this.getName(l), "TimeOfUseTariff.Awattar",
-							JsonUtils.buildJsonObject() //
-									.build())//
-			);
-			return new AppConfiguration(components,
-					Lists.newArrayList(ctrlEssTimeOfUseTariffDischargeId, "ctrlBalancing0"));
-		};
-	}
+			final var alias = this.getString(p, l, Property.ALIAS);
+			final var timeOfUseTariffProviderId = this.getId(t, p, Property.TIME_OF_USE_TARIF_PROVIDER_ID);
```
Comment (michaelgrill) [inline io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/AwattarHourly.java:103]: i usually separate the ids from the other variables and also put them at the top of the method

File: io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/TimeOfUseProps.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import io.openems.edge.core.appmanager.Nameable;
+import io.openems.edge.core.appmanager.OpenemsApp;
+import io.openems.edge.core.appmanager.Type.Parameter.BundleParameter;
+
+public final class TimeOfUseProps {
+
+	private TimeOfUseProps() {
+	}
+
+	private static final AppDef<OpenemsApp, Nameable, BundleParameter> defaultDef() {
```
Comment (michaelgrill) [inline io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/TimeOfUseProps.java:24]: you dont need a copy of this you can directly use `CommonProps.defaultDef()`

File: io.openems.edge.timeofusetariff.api/src/io/openems/edge/timeofusetariff/test/DummyTimeOfUseTariffProvider.java [hunk 4]
```diff
@@ -18,13 +18,31 @@ public class DummyTimeOfUseTariffProvider implements TimeOfUseTariff {
...
+		var quarterlyPrices = new Float[24];
 
 		for (var i = 0; i < 24; i++) {
 			quarterlyPrices[i] = hourlyPrices[i];
-			quarterlyPrices[i + 1] = hourlyPrices[i];
-			quarterlyPrices[i + 2] = hourlyPrices[i];
-			quarterlyPrices[i + 3] = hourlyPrices[i];
+		}
+
+		return new DummyTimeOfUseTariffProvider(now, quarterlyPrices);
```
Comment (michaelgrill) [inline io.openems.edge.timeofusetariff.api/src/io/openems/edge/timeofusetariff/test/DummyTimeOfUseTariffProvider.java:27]: why dont you directly return `new DummyTimeOfUseTariffProvider(now, hourlyPrices);`?

**Review by michaelgrill — CHANGES_REQUESTED** [review]

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+	 * 
+	 * @param expensivePeriodIndex The period index before which the cheap hour
+	 *                             should be found.
+	 * @param periods              The list of periods.
+	 * @param essUsableEnergy      The usable energy in the battery.
+	 * @return The index of the cheapest hour, or null if none is found.
+	 */
+	private Integer getCheapestHourBeforePeriod(int expensivePeriodIndex, List<Period> periods, int essUsableEnergy) {
+
+		Integer cheapHourIndex = null;
```
[the remark this answers is not in the document]
Comment (venu-sagar) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java:316]: Nothing concrete, just thought to have a null check if null is returned. 
Changing it to return -1.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+	 * @param cheapHour index of the cheap period.
+	 */
+	private void updateSchedule(int energy, int cheapHour) {
+		var period = this.periods.get(cheapHour);
+
+		switch (this.controlMode) {
+		case CHARGE_CONSUMPTION:
+			energy = -energy; // Charge energy is always negative.
+
+			if (period.chargeDischargeEnergy <= 0) {
```
[the remark this answers is not in the document]
Comment (venu-sagar) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java:429]: I understand the concern. Value cannot be null here. I handled that in the initial simulation itself. So at the end schedule initialization itself, null values are replaced.

The alternate way is to initialize with 0, but the problem is that Energy variables with 0 are still valid. It's hard to find whether it is from initialization or from the simulation calculations.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java:316]: In fact I believe returning `null` if no "cheapest period" was found, is more intuitive for a Java developer than returning `-1`.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+	 */
+	private CompletableFuture<? extends JsonrpcResponseSuccess> handleGetScheduleRequest(User user,
+			GetScheduleRequest request) {
+
+		var now = TimeOfUseTariffUtils.getNowRoundedDownToMinutes(this.componentManager.getClock(), MINUTES_PER_PERIOD);
+		var fromDate = now.minusHours(3);
+
+		var channeladdressPrices = new ChannelAddress("ctrlEssTimeOfUseTariff0", "QuarterlyPrices");
+		var channeladdressStateMachine = new ChannelAddress("ctrlEssTimeOfUseTariff0", "StateMachine");
+		var channeladdress = Stream.of(channeladdressPrices, channeladdressStateMachine).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:392]: @venu-sagar You misunderstood Michael. You can simply write `Set.of(channeladdressPrices, channeladdressStateMachine)`

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+	 * @param request the GetScheduleRequest
+	 * @return the Future JSON-RPC Response
+	 */
+	private CompletableFuture<? extends JsonrpcResponseSuccess> handleGetScheduleRequest(User user,
+			GetScheduleRequest request) {
+
+		var now = TimeOfUseTariffUtils.getNowRoundedDownToMinutes(this.componentManager.getClock(), MINUTES_PER_PERIOD);
+		var fromDate = now.minusHours(3);
+
+		var channeladdressPrices = new ChannelAddress("ctrlEssTimeOfUseTariff0", "QuarterlyPrices");
```
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:390]: @venu-sagar Why don't you use `this.id()` here, instead of hard-coded `ctrlEssTimeOfUseTariff0`. Won't this break if somebody configures a different ID?

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (venu-sagar) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:392]: > @venu-sagar You misunderstood Michael. You can simply write `Set. of(channeladdressPrices, channeladdressStateMachine)`

Yes, I understood that. it was silly that I complicated it with Stream. I made those changes locally. I need to test all my changes before I push them to the public.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package io.openems.edge.controller.ess.timeofusetariff;
+
+import io.openems.common.types.OptionsEnum;
+
+public enum ControlMode implements OptionsEnum {
+	CHARGE_CONSUMPTION(0, "Charge consumption from from grid"), //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java:6]: from from

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package io.openems.edge.controller.ess.timeofusetariff;
+
+import io.openems.common.types.OptionsEnum;
+
+public enum ControlMode implements OptionsEnum {
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java:5]: Why is this a `OptionsEnum`? For config settings a normal enum is better suitable,

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	CHARGE_CONSUMPTION(0, "Charge consumption from from grid"), //
+	DELAY_DISCHARGE(1, "Dealys discharge for high price hours"); //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java:7]: "Delays"; how about "Delays discharge during low-price hours?"

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	@Override
+	public String getName() {
+		return this.name;
+	}
+
+	@Override
+	public OptionsEnum getUndefined() {
+		return CHARGE_CONSUMPTION;
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java:29]: Non-intuitive

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java [hunk 13]
```diff
@@ -0,0 +1,467 @@
+package io.openems.edge.controller.ess.timeofusetariff;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import io.openems.edge.common.type.TypeUtils;
+
+public class Schedule {
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java:10]: I like the implementation of this class. It's clear and concise! 👍

I did not check the entire logic, but I assume both of you checked it - right?

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+			var period = this.periods.get(index);
+
+			// calculate initial energy for current period.
+			final int essInitialEnergyForCurrentPeriod = TypeUtils.max(0, essInitialEnergy);
+
+			// Update the value.
+			period.essInitialEnergy = essInitialEnergyForCurrentPeriod;
+
+			// updates 'chargeDischarge energy' and 'Grid energy' values.
+			this.updateEssAndGridEnergyInPeriod(index);
```
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java:197]: What's wrong with Michaels suggestion? You don't need 'index' in `updateEssAndGridEnergyInPeriod()`, so passing period seems to be better. 

(It's always a good idea to have strong typing (i.e. Period) instead of generic types like an int)

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/StateMachine.java [hunk 15]
```diff
@@ -1,12 +1,13 @@
...
 
 import io.openems.common.types.OptionsEnum;
 
 public enum StateMachine implements OptionsEnum {
 
 	DELAYED(0, "Delayed"), //
 	ALLOWS_DISCHARGE(1, "No active limitation, discharge is allowed"), //
-	STANDBY(2, "Outside controller time limits"); //
+	STANDBY(2, "Excess PV available."), //
+	CHARGING(3, "Charging the battery from grid."); //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/StateMachine.java:10]: I suggest: `DELAY_DISCHARGE`, `ALLOW_DISCHARGE`, `CHARGE_FROM_PV`, `CHARGE_FROM_GRID` with appropriate descriptions.

All names without final fullstop please.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java [hunk 16]
```diff
@@ -9,58 +9,63 @@
...
 				.text("Current state of the Controller")),
 
+		/**
+		 * Aggregated seconds when storage is being force charged.
+		 */
+		CHARGED_TIME(Doc.of(OpenemsType.LONG) //
+				.unit(Unit.CUMULATED_SECONDS) //
+				.persistencePriority(PersistencePriority.HIGH)), //
+		CHARGED(Doc.of(OpenemsType.BOOLEAN)//
+				.text("The controller currently recharge the battery")),
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java:29]: `CHARGING` would be more correct here... but for backwards compatibility to be similar with `DELAYED` I guess it is fine.

The text is grammatically wrong

Isn't the STATE_MACHINE sufficient to serve all these `CHARGED`, `DELAYED`, etc. channels?

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java [hunk 17]
```diff
@@ -9,58 +9,63 @@
...
+				.text("The amount of energy set to charge from grid in a period")), //
+
 		/**
 		 * Aggregated seconds when storage is blocked for discharge.
 		 */
 		DELAYED_TIME(Doc.of(OpenemsType.LONG) //
 				.unit(Unit.CUMULATED_SECONDS) //
 				.persistencePriority(PersistencePriority.HIGH)), //
-
 		DELAYED(Doc.of(OpenemsType.BOOLEAN)//
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java:40]: Should be DELAYING then... but can be kept for backwards compatibility.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java [hunk 18]
```diff
@@ -9,58 +9,63 @@
...
-		REMAINING_CONSUMPTION(Doc.of(OpenemsType.DOUBLE) //
-				.text("remaining consmption to charge from grid")),
-		TARGET_HOURS(Doc.of(OpenemsType.INTEGER) //
-				.text("Number of Target Hours")),
+		DISCHARGE_VALUE(Doc.of(OpenemsType.INTEGER) //
+				.unit(Unit.WATT)//
+				.text("The amount of energy set to discharge from the battery in a period")), //
+
+		CHARGE_DISCHARGE_ENERGY(Doc.of(OpenemsType.INTEGER) //
+				.text("Charge/Discharge energy calculated for the period.")), //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java:47]: So this is the same as CHARGE_VALUE and DISCHARGE_VALUE?

Please remove unnecessary channels. 

Be aware that any Channel that is required for UI has to have a persistencePriority of HIGH.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java [hunk 19]
```diff
@@ -9,58 +9,63 @@
...
 				.text("Predicted Consumption for the current quarterly hour")),
 		MIN_SOC(Doc.of(OpenemsType.INTEGER) //
-				.text("Minimum SoC to avoid complete discharge")),
-		PREDICTED_SOC_WITHOUT_LOGIC(Doc.of(OpenemsType.INTEGER) //
-				.text("SoC prediction curve without controller logic")),;
+				.text("Minimum SoC to avoid complete discharge")), //
+		QUARTERLY_PRICES(Doc.of(OpenemsType.FLOAT) //
+				.unit(Unit.EUROS_PER_MEGAWATT_HOUR) //
+				.text("Price of the electricity for the current Hour")
+				.persistencePriority(PersistencePriority.VERY_HIGH)), //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java:64]: Why is this VERY_HIGH?

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java [hunk 20]
```diff
@@ -9,58 +9,63 @@
...
 		PREDICTED_PRODUCTION(Doc.of(OpenemsType.INTEGER) //
 				.text("Predicted Production for the current quarterly hour")),
 		PREDICTED_CONSUMPTION(Doc.of(OpenemsType.INTEGER) //
 				.text("Predicted Consumption for the current quarterly hour")),
 		MIN_SOC(Doc.of(OpenemsType.INTEGER) //
-				.text("Minimum SoC to avoid complete discharge")),
-		PREDICTED_SOC_WITHOUT_LOGIC(Doc.of(OpenemsType.INTEGER) //
-				.text("SoC prediction curve without controller logic")),;
+				.text("Minimum SoC to avoid complete discharge")), //
+		QUARTERLY_PRICES(Doc.of(OpenemsType.FLOAT) //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffController.java:61]: Should be `QUARTERLY_PRICE`, because it holds only one price

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+	@Reference
+	private PredictorManager predictorManager;
+
+	@Reference
+	private TimeOfUseTariff timeOfUseTariff;
+
+	private Config config = null;
+	private Schedule schedule;
+	private ZonedDateTime nextQuarter = null;
+	private static final int PERIODS_PER_HOUR = 4;
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:105]: Please move statics to the top

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+	 * @param netEssCapacity net capacity of the battery.
+	 * @return the amount of energy that is limited.
+	 */
+	private int getLimitEnergy(int netEssCapacity) {
+
+		// Usable capacity based on minimum SoC from Limit total discharge and emergency
+		// reserve controllers.
+		var limitSoc = IntStream.concat(//
+				this.ctrlLimitTotalDischarges.stream().mapToInt(ctrl -> ctrl.getMinSoc().get()), //
+				this.ctrlEmergencyCapacityReserves.stream().mapToInt(ctrl -> ctrl.getActualReserveSoc().get())) //
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:306]: Possible NullPointerExceptions?

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+
+	/**
+	 * Force charges or delays the discharge if the schedule is set for current
+	 * period.
+	 * 
+	 * @throws OpenemsNamedException on error.
+	 */
+	private void setChargeOrDischarge() throws OpenemsNamedException {
+		if (this.schedule.periods.isEmpty()) {
+			return;
```
Comment (sfeilmeier) [inline io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:248]: If this method throws an Exception or returns here, the channels will not get set.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java [hunk 24]
```diff
@@ -0,0 +1,477 @@
...
+			var state = states.get(index);
+
+			if (price.isJsonNull() || state.isJsonNull()) {
+				continue;
+			}
+
+			result.add("timestamp",
+					new JsonPrimitive(timestamp.plusMinutes(15 * index).format(DateTimeFormatter.ISO_INSTANT)));
+			result.add("price", prices.get(index));
+			result.add("state", states.get(index));
```
[the remark this answers is not in the document]
Comment (sfeilmeier) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java:462]: I agree. I suggest introducing a `record` or separate class that handles the data and executes this quite complex logic. Currently all of this is not JUnit tested, right?

File: io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/AwattarHourly.java [hunk 25]
```diff
@@ -48,15 +52,41 @@
...
  */
 @org.osgi.service.component.annotations.Component(name = "App.TimeOfUseTariff.Awattar")
-public class AwattarHourly extends AbstractEnumOpenemsApp<Property> implements OpenemsApp {
+public class AwattarHourly extends AbstractOpenemsAppWithProps<AwattarHourly, Property, Type.Parameter.BundleParameter>
+		implements OpenemsApp {
+
+	public static enum Property implements Type<Property, AwattarHourly, Type.Parameter.BundleParameter>, Nameable {
+		// Components
+		CTRL_ESS_TIME_OF_USE_TARIF_ID(AppDef.componentId("ctrlEssTimeOfUseTariff0")), //
+		TIME_OF_USE_TARIF_PROVIDER_ID(AppDef.componentId("timeOfUseTariff0")), //
```
Comment (sfeilmeier) [inline io.openems.edge.core/src/io/openems/edge/app/timeofusetariff/AwattarHourly.java:61]: `TARIFF` with double "F" in both lines

File: io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties [hunk 26]
```diff
@@ -164,6 +164,10 @@ App.PvInverter.Sma.phase.description = Angeschlossene Phase(n) des Wechselrichte
 App.PvInverter.SolarEdge.Name = SolarEdge PV-Wechselrichter
 
 # Time of use Tarif
+App.TimeOfUseTariff.controlMode.label = Kontroller Modus
```
Comment (sfeilmeier) [inline io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties:167]: "Kontroller"?

File: io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties [hunk 27]
```diff
@@ -164,6 +164,10 @@ App.PvInverter.Sma.phase.description = Angeschlossene Phase(n) des Wechselrichte
...
+App.TimeOfUseTariff.controlMode.chargeConsumption = Aufladen der Batterie zu günstigen Preis
```
Comment (sfeilmeier) [inline io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties:168]: "günstigem"

File: io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties [hunk 28]
```diff
@@ -164,6 +164,10 @@ App.PvInverter.Sma.phase.description = Angeschlossene Phase(n) des Wechselrichte
...
+App.TimeOfUseTariff.controlMode.delayDischarge = Verzögernung der Entladung während günstigen Preisen
```
Comment (sfeilmeier) [inline io.openems.edge.core/src/io/openems/edge/core/appmanager/translation_de.properties:169]: "Verzögerung", "günstiger Preise"


Please validate texts with a native German speaker (i.e. not Michael 😉 but Jan)

File: io.openems.edge.timeofusetariff.awattar/src/io/openems/edge/timeofusetariff/awattar/TimeOfUseTariffAwattarImpl.java [hunk 29]
```diff
@@ -115,16 +115,11 @@ protected void deactivate() {
...
 		 */
 		var now = ZonedDateTime.now();
-		var nextRun = now.withHour(14).truncatedTo(ChronoUnit.HOURS);
-		if (now.isAfter(nextRun)) {
-			nextRun = nextRun.plusDays(1);
-		}
-
-		var duration = Duration.between(now, nextRun);
-		var delay = duration.getSeconds();
+		var nextRun = now.plusHours(1).truncatedTo(ChronoUnit.HOURS);
```
Comment (sfeilmeier) [inline io.openems.edge.timeofusetariff.awattar/src/io/openems/edge/timeofusetariff/awattar/TimeOfUseTariffAwattarImpl.java:121]: Please add a comment why this is now needed every hour. From what I understood aWATTar does not anymore provide full day-ahead prices, but only 24 hours ahead, right?

**Review by sfeilmeier — CHANGES_REQUESTED** [review]

**Comment by sfeilmeier** [comment]
One more question: we do not have a Risk Level anymore? It seemed to be a good idea before.

File: io.openems.edge.controller.ess.timeofusetariff/readme.adoc [hunk 30]
```diff
@@ -0,0 +1,11 @@
+= ESS Time-of-Use Tariff
+
+This Controller optimizes the performance of an energy storage system (ESS) in conjunction with a Time-Of-Use (ToU) Tariff provider. 
+
+It operates every 15 minutes, setting the ESS schedule for the remaining day in 15-minute intervals.
```
[the remark this answers is not in the document]
Comment (venu-sagar) [reply io.openems.edge.controller.ess.timeofusetariff/readme.adoc:5]: The prices undergo daily updates at 2:00 pm, resulting in the availability of 96 values by that time. However, by 1:00 pm, the values are only available until midnight. Consequently, we schedule ESS exclusively until midnight in such scenarios, highlighting that a 24-hour span is not always applicable.

File: io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java (hunk 11 above)
Comment (venu-sagar) [reply io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/ControlMode.java:7]: Makes sense.
Anyways it should have been "Delays discharge to be usable during high-price hours", your suggestion is easier to understand.

**Comment by sfeilmeier** [comment]
@venu-sagar Can we please re-add the "Risk Level" again? Please discuss with Jan on an easy-to-implement short-term solution for this. We should not drop the config option, as discussed.

**Comment by venu-sagar** [comment]
> @venu-sagar Can we please re-add the "Risk Level" again? Please discuss with Jan on an easy-to-implement short-term solution for this. We should not drop the config option, as discussed.

Added the "Risk Level" to the controller.

**Review by sfeilmeier — APPROVED** [review]
